clients/drcachesim/reader/file_reader.h

```diff
@@ -107,287 +95,83 @@ template <typename T> class file_reader_t : public reader_t {
 
-        thread_count_ = input_files_.size();
-        queues_.resize(input_files_.size());
-        tids_.resize(input_files_.size());
-        timestamps_.resize(input_files_.size());
-        times_.resize(input_files_.size(), 0);
-        // We can't take the address of a vector<bool> element so we use a raw array.
-        thread_eof_ = new bool[input_files_.size()];
-        memset(thread_eof_, 0, input_files_.size() * sizeof(*thread_eof_));
-
-        // First read the tid and pid entries which precede any timestamps_.
+        // First read the tid and pid entries which precede any timestamps.
         // We hand out the tid to the output on every thread switch, and the pid
         // the very first time for the thread.
-        trace_entry_t header, next, pid = {};
-        for (index_ = 0; index_ < input_files_.size(); ++index_) {
-            if (!read_next_thread_entry(index_, &header, &thread_eof_[index_]) ||
-                header.type != TRACE_TYPE_HEADER) {
-                ERRMSG("Invalid header for input file #%zu\n", index_);
-                return false;
-            }
-            // We can handle the older version 1 as well which simply omits the
-            // early marker with the arch tag, and version 2 which only differs wrt
-            // TRACE_MARKER_TYPE_KERNEL_EVENT..
-            if (header.addr > TRACE_ENTRY_VERSION) {
-                ERRMSG(
-                    "Cannot handle version #%zu (expect version <= #%u) for input file "
-                    "#%zu\n",
-                    header.addr, TRACE_ENTRY_VERSION, index_);
+        trace_entry_t *entry;
+        trace_entry_t header = {}, pid = {}, tid = {};
+        entry = read_next_entry();
+        if (entry == nullptr || entry->type != TRACE_TYPE_HEADER) {
+            ERRMSG("Invalid header\n");
+            return false;
+        }
+        header = *entry;
+        // We can handle the older version 1 as well which simply omits the
+        // early marker with the arch tag, and version 2 which only differs wrt
+        // TRACE_MARKER_TYPE_KERNEL_EVENT..
+        if (entry->addr > TRACE_ENTRY_VERSION) {
+            ERRMSG("Cannot handle version #%zu (expect version <= #%u)\n", entry->addr,
+                   TRACE_ENTRY_VERSION);
+            return false;
+        }
+        // Read the meta entries until we hit the pid.
+        // We want to pass the tid+pid to the reader *before* any markers,
+        // even though 2 markers preced the tid+pid in the file.
+        std::queue<trace_entry_t> marker_queue;
+        while ((entry = read_next_entry()) != nullptr) {
+            if (entry->type == TRACE_TYPE_PID) {
+                // We assume the pid entry is the last, right before the timestamp.
+                pid = *entry;
+                break;
+            } else if (entry->type == TRACE_TYPE_THREAD)
+                tid = *entry;
+            else if (entry->type == TRACE_TYPE_MARKER)
+                marker_queue.push(*entry);
```

Comment:
Add to comment which markers we expect here. Looking at `offline_instr_t::append_thread_header` I see only the OFFLINE_EXT_TYPE_HEADER entry before the tid one but no marker.

Reply:
It is legacy traces which have something different: I think they used to have the timestamp before the tid+pid in the raw file.  Updated the comment.